dbt_common/behavior_flags.py

```diff
@@ -0,0 +1,108 @@
+import inspect
+from types import SimpleNamespace
+from typing import Any, Dict, List, TypedDict
+
+try:
+    from typing import NotRequired
+except ImportError:
+    # NotRequired was introduced in Python 3.11
+    # This is the suggested way to implement a TypedDict with optional arguments
+    from typing import Optional as NotRequired
+
+from dbt_common.events.base_types import WarnLevel
+from dbt_common.events.functions import fire_event
+from dbt_common.events.types import BehaviorDeprecationEvent
+
+
+class BehaviorFlag:
+    """
+    The canonical behavior flag that gets used through dbt packages
+
+    Args:
+        name: the name of the behavior flag, e.g. enforce_quoting_on_relation_creation
+        setting: the flag setting, after accounting for user input and the default
+        deprecation_event: the event to fire if the flag evaluates to False
+    """
+
+    def __init__(self, name: str, setting: bool, deprecation_event: WarnLevel) -> None:
+        self.name = name
+        self.setting = setting
+        self.deprecation_event = deprecation_event
+
+    @property
+    def setting(self) -> bool:
+        if self._setting is False:
+            fire_event(self.deprecation_event)
+        return self._setting
+
+    @setting.setter
+    def setting(self, value: bool) -> None:
+        self._setting = value
+
+    def __bool__(self) -> bool:
+        return self.setting
+
+
+class RawBehaviorFlag(TypedDict):
+    """
+    A set of config used to create a BehaviorFlag
+
+    Args:
+        name: the name of the behavior flag
+        default:  default setting, starts as False, becomes True in the next minor release
+        deprecation_version: the version when the default will change to True
+        deprecation_message: an additional message to send when the flag evaluates to False
+        docs_url: the url to the relevant docs on docs.getdbt.com
+    """
+
+    name: str
+    default: bool
+    source: NotRequired[str]
+    deprecation_version: NotRequired[str]
+    deprecation_message: NotRequired[str]
+    docs_url: NotRequired[str]
+
+
+# this is effectively a dictionary that supports dot notation
+# it makes usage easy, e.g. adapter.behavior.my_flag
+Behavior = SimpleNamespace
+
+
+def register(
+    behavior_flags: List[RawBehaviorFlag],
+    user_flags: Dict[str, Any],
```

Comment:
A name like 'user_settings' or 'user_overrides' might better express the intention. Also, I think that Dist[str, bool] is the right type to use here, if we are only allowing true/false.

Reply:
I'm fine with `user_overrides`. The reason I had `Dict[str, Any]` was because I'm passing in the entire `flags` collection and I wasn't sure what the rest of that looked like. It would make sense that something named `flags` was a collection of `bool`s, but I didn't want to assume.